crates/attestation/src/crypto/crl.rs

```diff
@@ -0,0 +1,433 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use crate::crypto::{SubjectPublicKeyInfoExt, TbsCertificateExt};
+
+use anyhow::{bail, ensure, Context, Result};
+use der::{Encode, Sequence};
+use x509::crl::CertificateList;
+use x509::name::Name;
+use x509::PkiPath;
+
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct CrlListEntry<'a> {
+    pub url: String,
+    pub crl: CertificateList<'a>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct CrlList<'a> {
+    pub crls: Vec<CrlListEntry<'a>>,
+}
+
+impl<'a> CrlList<'a> {
+    fn get_crl_by_url(&self, url: &str) -> Option<&CertificateList> {
+        for pair in &self.crls {
+            if pair.url == url {
+                return Some(&pair.crl);
+            }
+        }
+        None
+    }
+
+    fn get_crl_by_issuer(&self, issuer: &Name) -> Option<&CertificateList> {
+        for pair in &self.crls {
+            if &pair.crl.tbs_cert_list.issuer == issuer {
+                return Some(&pair.crl);
+            }
+        }
+        None
+    }
+}
+
+pub trait PkiPathCRLCheck<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()>;
+}
+
+impl<'a> PkiPathCRLCheck<'a> for PkiPath<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()> {
+        // We want to ensure that a valid CRL was passed, so we make sure at least one of the CRLs
+        // is valid for this `PkiPath`, otherwise, no valid CRLs were received, and that's not okay.
+        let mut found = false;
+        let mut iter = self.iter().peekable();
+        loop {
+            let cert = iter.next();
+            let next = iter.peek();
+            if cert.is_none() || next.is_none() {
+                break;
+            }
+            let cert = cert.unwrap();
+            let next = next.unwrap();
+
+            let crls = {
+                let mut crls = vec![];
+                if let Some(crl) = pairs.get_crl_by_issuer(&cert.tbs_certificate.subject) {
```

Comment:
I think we should not look up by issuer name, and just always look up by the URL specified in the cert.

Reply:
Agreed, but I was having a problem where looking up by URL for the Intel certs was returning the wrong CRL (despite having the correct URL in the new CRL file format). This will be fixed in the follow-up PR.

Reply:
Okay, in what way do you mean that?
i.e. what CRL do you get, which do you expect?

Reply:
Looking up the CRL from the Intel SGX Processor cert ended up actually retrieving the CRL for the root cert, and then failed. That's why I had all the debugging print statements, so see that the data in the file was correct, yet wrong CRL bring returned. I has a few posts on Rocket chat about it https://chat.enarx.dev/channel/development?msg=eYp8C6sybwSuimGfo.

Reply:
```
#2: CA Cert: CN=Intel SGX PCK Processor CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        CRL URL: ["https://certificates.trustedservices.intel.com/IntelSGXRootCA.der"], Issuer: CN=Intel SGX Root CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
CRL validation error: signature error
        CA Cert algo: 1.2.840.10045.4.3.2
        CA Cert subject: CN=Intel SGX PCK Processor CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        CA Cert issuer: CN=Intel SGX Root CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        Next Cert algo: 1.2.840.10045.4.3.2
        Next Cert subject: CN=Intel SGX PCK Certificate,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        Next Cert issuer: CN=Intel SGX PCK Processor CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        CRL issuer: CN=Intel SGX Root CA,O=Intel Corporation,L=Santa Clara,STATEORPROVINCENAME=CA,C=US
        CRL algo: 1.2.840.10045.4.3.2
        URL: ["https://certificates.trustedservices.intel.com/IntelSGXRootCA.der"]
```

Reply:
Okay. Let's merge it as-is then and look at it soon.